File: javascript/087.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var isScramble = function(s1, s2) {
  if (s1.length !== s2.length) return false;

  let s1Map = {};
  for(let i = 0; i < s1.length; ++i) {
    const c = s1[i];
    c in s1Map ?  s1Map[c] += 1 : s1Map[c] = 0;
  }

  let s2Map = {};
  for(let i = 0; i < s2.length; ++i) {
    const c = s2[i];
    c in s2Map ?  s2Map[c] += 1 : s2Map[c] = 0;
  }

  for(key in s1Map) {
    if (s1Map[key] !== s2Map[key]) return false;
  }

  if (s1.length === 1) return s1 === s2;

  for(let i = 1; i < s1.length; ++i) {
    const s1Part1 = s1.substr(0, i);
    const s1Part2 = s1.substr(i);
    const s2Part11 = s2.substr(0, i);
    const s2Part21 = s2.substr(i);
    const s2Part12 = s2.substr(0, s2.length - i);
    const s2Part22 = s2.substr(s2.length - i);
    if (isScramble(s1Part1, s2Part11) && isScramble(s1Part2, s2Part21)) { return true; }
    if (isScramble(s1Part1, s2Part22) && isScramble(s1Part2, s2Part12)) { return true; }
  }

  return false;
};

File: ruby/087.rb
# @param {String} s1
# @param {String} s2
# @return {Boolean}
def is_scramble(s1, s2)
  return false if s1.length != s2.length
  return s1 == s2 if s1.length == 1

  all_match = true
  s1_array = s1.split("").sort!
  s2_array = s2.split("").sort!
  s1_array.each_with_index do |char, index|
    if char != s2_array[index]
      all_match = false
      break
    end
  end
  return false if !all_match

  any_match = false
  1.upto(s1.length - 1).each do |end_pos|
    s1_part1 = s1[0...end_pos]
    s1_part2 = s1[end_pos..s1.length]

    s2_part1 = s2[0...end_pos]
    s2_part2 = s2[end_pos..s2.length]
    if is_scramble(s1_part1, s2_part1) && is_scramble(s1_part2, s2_part2)
      any_match = true
      break
    end

    end_pos = s1.length - end_pos
    s2_part1 = s2[0...end_pos]
    s2_part2 = s2[end_pos...s2.length]
    if is_scramble(s1_part1, s2_part2) && is_scramble(s1_part2, s2_part1)
      any_match = true
      break
    end
  end

  return any_match
end